File: app/routers/ai.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app import models, schemas
from app.auth import get_db, get_current_user
from app.core.ai.context import build_ai_context
from app.core.ai.planning import recommend_team_plan
from app.core.ai.rules import build_chat_fallback_reply, build_conflicts, build_insights, fallback_summary
from app.core.ai.gemini import generate_grounded_summary, gemini_configured, generate_chat_reply

router = APIRouter(prefix="/ai", tags=["ai"])


@router.get(
    "/insights",
    response_model=schemas.AIInsightsResponse,
    summary="AI insights",
    description="Returns role-aware AI insights grounded in current projects, tasks, workload, and comments.",
)
def ai_insights(
    db: Session = Depends(get_db),
    current: models.User = Depends(get_current_user),
):
    context = build_ai_context(db, current)
    conflicts = build_conflicts(context)
    insights = build_insights(context, conflicts)
    summary, source = fallback_summary(current.role.value, insights, conflicts)
    if gemini_configured():
        gemini_summary = generate_grounded_summary(current.role.value, context, insights, conflicts)
        if gemini_summary:
            summary = gemini_summary
            source = "gemini"
    return schemas.AIInsightsResponse(
        summary=summary,
        source=source,
        generated_at=datetime.utcnow(),
        scope=current.role.value,
        insights=[schemas.AIInsightItem(**item) for item in insights],
    )


@router.get(
    "/workload-conflicts",
    response_model=schemas.AIConflictsResponse,
    summary="AI conflict detection",
    description="Returns workload and execution conflicts for the current role scope.",
)
def ai_workload_conflicts(
    db: Session = Depends(get_db),
    current: models.User = Depends(get_current_user),
):
    context = build_ai_context(db, current)
    conflicts = build_conflicts(context)
    insights = build_insights(context, conflicts)
    summary, source = fallback_summary(current.role.value, insights, conflicts)
    if gemini_configured():
        gemini_summary = generate_grounded_summary(current.role.value, context, insights, conflicts)
        if gemini_summary:
            summary = gemini_summary
            source = "gemini"
    return schemas.AIConflictsResponse(
        summary=summary,
        source=source,
        generated_at=datetime.utcnow(),
        scope=current.role.value,
        conflicts=[schemas.AIConflictItem(**item) for item in conflicts],
    )


@router.post(
    "/chat",
    response_model=schemas.AIChatResponse,
    summary="AI chat",
    description="Accepts a user message and optional recent history. Returns a grounded AI reply for the current role scope.",
)
def ai_chat(
    payload: schemas.AIChatRequest,
    db: Session = Depends(get_db),
    current: models.User = Depends(get_current_user),
):
    context = build_ai_context(db, current)
    history = [item.model_dump() for item in payload.history]
    reply = generate_chat_reply(current.role.value, context, payload.message, history)
    source = "gemini" if reply else "rules"
    if not reply:
        reply = build_chat_fallback_reply(payload.message, context)
    return schemas.AIChatResponse(
        reply=reply,
        source=source,
        generated_at=datetime.utcnow(),
        scope=current.role.value,
    )


@router.post(
    "/project-plan",
    response_model=schemas.AIProjectPlanResponse,
    summary="AI project planning",
    description="Accepts a project brief and returns recommended teams, members, and a starter task plan.",
)
def ai_project_plan(
    payload: schemas.AIProjectPlanRequest,
    db: Session = Depends(get_db),
    current: models.User = Depends(get_current_user),
):
    if current.role == models.Role.USER:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Forbidden")
    recommended_teams, tasks, recommended_people, suggested_team_size, suggested_roles = recommend_team_plan(
        db,
        payload.name,
        payload.description,
    )
    summary = (
        f"Built a {suggested_team_size}-person draft squad for {payload.name}, led by {recommended_people[0]['full_name']}."
        if recommended_people
        else f"No strong historical staffing match was found for {payload.name}."
    )
    return schemas.AIProjectPlanResponse(
        summary=summary,
        source="rules",
        generated_at=datetime.utcnow(),
        suggested_team_size=suggested_team_size,
        recommended_people=[schemas.AIRecommendedMember(**item) for item in recommended_people],
        suggested_roles=[schemas.AIRecommendedRoleSlot(**item) for item in suggested_roles],
        recommended_teams=[schemas.AIRecommendedTeam(**item) for item in recommended_teams],
        tasks=[schemas.AIPlannedTask(**item) for item in tasks],
    )

File: arsii_mvp/lib/models/ai.dart
class AIInsightItem {
  final String title;
  final String detail;
  final String priority;
  final String? entityType;
  final int? entityId;

  AIInsightItem({
    required this.title,
    required this.detail,
    required this.priority,
    required this.entityType,
    required this.entityId,
  });

  factory AIInsightItem.fromJson(Map<String, dynamic> json) {
    return AIInsightItem(
      title: json['title'] as String,
      detail: json['detail'] as String,
      priority: json['priority'] as String,
      entityType: json['entity_type'] as String?,
      entityId: json['entity_id'] as int?,
    );
  }
}

class AIInsightsResponse {
  final String summary;
  final String source;
  final String generatedAt;
  final String scope;
  final List<AIInsightItem> insights;

  AIInsightsResponse({
    required this.summary,
    required this.source,
    required this.generatedAt,
    required this.scope,
    required this.insights,
  });

  factory AIInsightsResponse.fromJson(Map<String, dynamic> json) {
    return AIInsightsResponse(
      summary: json['summary'] as String,
      source: json['source'] as String,
      generatedAt: json['generated_at'] as String,
      scope: json['scope'] as String,
      insights: (json['insights'] as List<dynamic>)
          .map((e) => AIInsightItem.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class AIConflictItem {
  final String title;
  final String detail;
  final String severity;
  final String recommendation;
  final String? entityType;
  final int? entityId;

  AIConflictItem({
    required this.title,
    required this.detail,
    required this.severity,
    required this.recommendation,
    required this.entityType,
    required this.entityId,
  });

  factory AIConflictItem.fromJson(Map<String, dynamic> json) {
    return AIConflictItem(
      title: json['title'] as String,
      detail: json['detail'] as String,
      severity: json['severity'] as String,
      recommendation: json['recommendation'] as String,
      entityType: json['entity_type'] as String?,
      entityId: json['entity_id'] as int?,
    );
  }
}

class AIConflictsResponse {
  final String summary;
  final String source;
  final String generatedAt;
  final String scope;
  final List<AIConflictItem> conflicts;

  AIConflictsResponse({
    required this.summary,
    required this.source,
    required this.generatedAt,
    required this.scope,
    required this.conflicts,
  });

  factory AIConflictsResponse.fromJson(Map<String, dynamic> json) {
    return AIConflictsResponse(
      summary: json['summary'] as String,
      source: json['source'] as String,
      generatedAt: json['generated_at'] as String,
      scope: json['scope'] as String,
      conflicts: (json['conflicts'] as List<dynamic>)
          .map((e) => AIConflictItem.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class AIChatMessage {
  final String role;
  final String content;

  AIChatMessage({
    required this.role,
    required this.content,
  });

  Map<String, dynamic> toJson() {
    return {
      'role': role,
      'content': content,
    };
  }
}

class AIChatResponse {
  final String reply;
  final String source;
  final String generatedAt;
  final String scope;

  AIChatResponse({
    required this.reply,
    required this.source,
    required this.generatedAt,
    required this.scope,
  });

  factory AIChatResponse.fromJson(Map<String, dynamic> json) {
    return AIChatResponse(
      reply: json['reply'] as String,
      source: json['source'] as String,
      generatedAt: json['generated_at'] as String,
      scope: json['scope'] as String,
    );
  }
}

class AIRecommendedMember {
  final int userId;
  final String fullName;
  final int? teamId;
  final String? teamName;
  final String? role;
  final String? availability;
  final int? matchPercentage;
  final String reason;

  AIRecommendedMember({
    required this.userId,
    required this.fullName,
    required this.teamId,
    required this.teamName,
    required this.role,
    required this.availability,
    required this.matchPercentage,
    required this.reason,
  });

  factory AIRecommendedMember.fromJson(Map<String, dynamic> json) {
    return AIRecommendedMember(
      userId: json['user_id'] as int,
      fullName: json['full_name'] as String,
      teamId: json['team_id'] as int?,
      teamName: json['team_name'] as String?,
      role: json['role'] as String?,
      availability: json['availability'] as String?,
      matchPercentage: json['match_percentage'] as int?,
      reason: json['reason'] as String,
    );
  }
}

class AIRecommendedRoleSlot {
  final String label;
  final int count;
  final String note;

  AIRecommendedRoleSlot({
    required this.label,
    required this.count,
    required this.note,
  });

  factory AIRecommendedRoleSlot.fromJson(Map<String, dynamic> json) {
    return AIRecommendedRoleSlot(
      label: json['label'] as String,
      count: json['count'] as int,
      note: json['note'] as String,
    );
  }
}

class AIRecommendedTeam {
  final int teamId;
  final String teamName;
  final double score;
  final String reason;
  final List<String> specialties;
  final List<AIRecommendedMember> members;

  AIRecommendedTeam({
    required this.teamId,
    required this.teamName,
    required this.score,
    required this.reason,
    required this.specialties,
    required this.members,
  });

  factory AIRecommendedTeam.fromJson(Map<String, dynamic> json) {
    return AIRecommendedTeam(
      teamId: json['team_id'] as int,
      teamName: json['team_name'] as String,
      score: (json['score'] as num).toDouble(),
      reason: json['reason'] as String,
      specialties: (json['specialties'] as List<dynamic>? ?? []).cast<String>(),
      members: (json['members'] as List<dynamic>)
          .map((e) => AIRecommendedMember.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

class AIPlannedTask {
  final String title;
  final String description;
  final int? recommendedAssigneeId;
  final String? recommendedAssigneeName;

  AIPlannedTask({
    required this.title,
    required this.description,
    required this.recommendedAssigneeId,
    required this.recommendedAssigneeName,
  });

  factory AIPlannedTask.fromJson(Map<String, dynamic> json) {
    return AIPlannedTask(
      title: json['title'] as String,
      description: json['description'] as String,
      recommendedAssigneeId: json['recommended_assignee_id'] as int?,
      recommendedAssigneeName: json['recommended_assignee_name'] as String?,
    );
  }
}

class AIProjectPlanResponse {
  final String summary;
  final String source;
  final String generatedAt;
  final int suggestedTeamSize;
  final List<AIRecommendedMember> recommendedPeople;
  final List<AIRecommendedRoleSlot> suggestedRoles;
  final List<AIRecommendedTeam> recommendedTeams;
  final List<AIPlannedTask> tasks;

  AIProjectPlanResponse({
    required this.summary,
    required this.source,
    required this.generatedAt,
    required this.suggestedTeamSize,
    required this.recommendedPeople,
    required this.suggestedRoles,
    required this.recommendedTeams,
    required this.tasks,
  });

  factory AIProjectPlanResponse.fromJson(Map<String, dynamic> json) {
    return AIProjectPlanResponse(
      summary: json['summary'] as String,
      source: json['source'] as String,
      generatedAt: json['generated_at'] as String,
      suggestedTeamSize: json['suggested_team_size'] as int,
      recommendedPeople: (json['recommended_people'] as List<dynamic>)
          .map((e) => AIRecommendedMember.fromJson(e as Map<String, dynamic>))
          .toList(),
      suggestedRoles: (json['suggested_roles'] as List<dynamic>)
          .map((e) => AIRecommendedRoleSlot.fromJson(e as Map<String, dynamic>))
          .toList(),
      recommendedTeams: (json['recommended_teams'] as List<dynamic>)
          .map((e) => AIRecommendedTeam.fromJson(e as Map<String, dynamic>))
          .toList(),
      tasks: (json['tasks'] as List<dynamic>)
          .map((e) => AIPlannedTask.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}
